File: algorithms/medium/non_decreasing_subsequences.cpp
class Solution {
public:
  vector<vector<int>> ans;
  vector<vector<int>> findSubsequences(vector<int> &nums) {
    vector<int> st;
    dfs(nums, 0, st);
    return ans;
  }
  void dfs(vector<int> &nums, int curr, vector<int> &st) {
    if (st.size() > 1) {
      ans.push_back(st);
    }
    unordered_set<int> mp;
    for (int i = curr; i < nums.size(); i++) {
      if (st.size() > 0 && nums[i] < st[st.size() - 1] || mp.find(nums[i]) != mp.end()) {
        continue;
      }
      mp.insert(nums[i]);
      st.push_back(nums[i]);
      dfs(nums, i + 1, st);
      st.pop_back();
    }
  }
};

File: algorithms/medium/non_decreasing_subsequences.ex
defmodule Solution do
  @spec find_subsequences(nums :: [integer]) :: [[integer]]
  def find_subsequences(nums, stack \\ [], ans \\ []) do
    ans = if length(stack) > 1, do: [Enum.reverse(stack) | ans], else: ans

    loop(nums, stack, ans, %MapSet{})
  end

  def loop([], _stack, ans, _set), do: ans

  def loop([n | ns], stack, ans, set) do
    cond do
      (stack != [] && n < hd(stack)) || n in set ->
        loop(ns, stack, ans, set)

      true ->
        loop(ns, stack, find_subsequences(ns, [n | stack], ans), MapSet.put(set, n))
    end
  end
end
